File: fcfs.js
#!/usr/bin/env node
/**
 * FCFS (First-Come First-Serve) - Non-preemptive scheduling
 */

function fcfs(processes) {
    // sort by arrival time (stable)
    const procs = [...processes].sort((a, b) => a.arrival - b.arrival);
    let time = 0;
    const result = [];
    
    for (const proc of procs) {
        const start = Math.max(time, proc.arrival);
        const completion = start + proc.burst;
        const turnaround = completion - proc.arrival;
        const waiting = start - proc.arrival;
        
        result.push({
            pid: proc.pid,
            arrival: proc.arrival,
            burst: proc.burst,
            start,
            completion,
            turnaround,
            waiting
        });
        
        time = completion;
    }
    
    return result;
}

function printTable(result) {
    console.log("PID".padStart(4), "AT".padStart(4), "BT".padStart(4), 
                "ST".padStart(4), "CT".padStart(4), "TAT".padStart(5), "WT".padStart(5));
    
    for (const row of result) {
        console.log(
            String(row.pid).padStart(4),
            String(row.arrival).padStart(4),
            String(row.burst).padStart(4),
            String(row.start).padStart(4),
            String(row.completion).padStart(4),
            String(row.turnaround).padStart(5),
            String(row.waiting).padStart(5)
        );
    }
    
    const n = result.length;
    const avgTAT = result.reduce((sum, r) => sum + r.turnaround, 0) / n;
    const avgWT = result.reduce((sum, r) => sum + r.waiting, 0) / n;
    
    console.log(`\nAverage Turnaround Time = ${avgTAT.toFixed(2)}`);
    console.log(`Average Waiting Time    = ${avgWT.toFixed(2)}`);
}

// Main execution
if (require.main === module) {
    const processes = [
        { pid: 1, arrival: 0, burst: 5 },
        { pid: 2, arrival: 2, burst: 3 },
        { pid: 3, arrival: 4, burst: 1 },
        { pid: 4, arrival: 6, burst: 2 }
    ];
    
    const result = fcfs(processes);
    console.log("FCFS Scheduling:\n");
    printTable(result);
}

module.exports = { fcfs, printTable };

File: fcfs.py
#!/usr/bin/env python3
"""Simple, short FCFS (First-Come First-Serve) scheduler example.
Input: list of (pid, arrival_time, burst_time).
Outputs: completion time, turnaround time, waiting time and averages.
"""

def fcfs(processes):
    # sort by arrival time (stable)
    procs = sorted(processes, key=lambda p: p[1])
    time = 0
    result = []
    for pid, arrival, burst in procs:
        start = max(time, arrival)
        completion = start + burst
        turnaround = completion - arrival
        waiting = start - arrival
        result.append((pid, arrival, burst, start, completion, turnaround, waiting))
        time = completion
    return result


def print_table(result):
    hdr = ("PID", "AT", "BT", "ST", "CT", "TAT", "WT")
    print("{:>4} {:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format(*hdr))
    for row in result:
        print("{:>4} {:>4} {:>4} {:>4} {:>4} {:>5} {:>5}".format(*row))
    n = len(result)
    avg_tat = sum(r[5] for r in result) / n
    avg_wt = sum(r[6] for r in result) / n
    print()
    print(f"Average Turnaround Time = {avg_tat:.2f}")
    print(f"Average Waiting Time    = {avg_wt:.2f}")


if __name__ == '__main__':
    # small example: list of (pid, arrival_time, burst_time)
    processes = [
        (1, 0, 5),
        (2, 2, 3),
        (3, 4, 1),
        (4, 6, 2),
    ]
    res = fcfs(processes)
    print_table(res)
